Add custom problem to dump action generated object syntax

dschwen opened the pull request:
Replacement for #8876 

Closes #8875

## Example

```
./tensor_mechanics-opt -i test/tests/action/two_block_new.i Problem/type=DumpObjectsProblem Problem/dump_path=Modules/TensorMechanics/Master/block2
```

will output

```
[AuxKernels]
  [./stress_xx_block2]
    type = RankTwoAux
    block = 2
    execute_on = TIMESTEP_END
    index_i = 0
    index_j = 0
    rank_two_tensor = stress
    variable = stress_xx
  [../]
  [./strain_yy_block2]
    type = RankTwoAux
    block = 2
    execute_on = TIMESTEP_END
    index_i = 1
    index_j = 1
    rank_two_tensor = total_strain
    variable = strain_yy
  [../]
[]

[AuxVariables]
  [./stress_xx]
    blocks = '1 2'
    family = MONOMIAL
    order = CONSTANT
  [../]
  [./strain_yy]
    blocks = '1 2'
    family = MONOMIAL
    order = CONSTANT
  [../]
[]

[Kernels]
  [./TM_block20]
    type = StressDivergenceTensors
    block = 2
    component = 0
    displacements = 'disp_x disp_y'
    use_displaced_mesh = true
    variable = disp_x
  [../]
  [./TM_block21]
    type = StressDivergenceTensors
    block = 2
    component = 1
    displacements = 'disp_x disp_y'
    use_displaced_mesh = true
    variable = disp_y
  [../]
[]

[Materials]
  [./block2_strain]
    type = ComputeFiniteStrain
    block = 2
    displacements = 'disp_x disp_y'
  [../]
[]

[Variables]
  [./disp_x]
    blocks = '1 2'
  [../]
  [./disp_y]
    blocks = '1 2'
  [../]
[]
```

([with a minimal set of input parameters](https://gist.github.com/dschwen/10c91443e1d88dcc5ed3fba050aa05f8/revisions)), which is what the

```
[Modules/TensorMechanics/Master]
  add_variables = true
  strain = FINITE
  generate_output = 'stress_xx'
  [./block2]
    block = 2
    additional_generate_output = 'strain_yy'
  [../]
[]
```

block in this input file creates.

dschwen committed:
Add custom problem to dump action generated object syntax (#8875)

dschwen committed:
Output bools as true/false (#8875)

dschwen committed:
Minimal parameter set, add variables (#8875)

dschwen committed:
Update (#8875)

permcody commented on framework/src/problems/DumpObjectsProblem.C:
```diff
@@ -0,0 +1,349 @@
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
```
Wrong header!

dschwen committed:
Address comments fix bug (#8875)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/176906/) on 8c66ea5 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11269/site)

This comment will be updated on new commits.

dschwen commented:
Guys, this is not adding syntax that will be used in input files. Hence if in 5 years you come up with a better solution it can easily be deprecated and removed. We wouldn't be breaking anyones stuff in the process.

permcody commented:
Tag @idaholab/moose-team 

I'm fine with this. It's working code and obviously in demand. It's certainly not a complete solution and we do have better ideas that will take some time. For those of you wanting this capability, let's just be aware that it will likely change substantially so I hope it doesn't get spread all over. Also, since it is bolted on in this design, it could fall behind and certainly won't expand more advanced sections without continued support. I'm not going to stand in the way of merging this, but I'm also not going to press the button either. Any other takers on the team?

andrsd reviewed (changes requested):
This needs tests

aeslaughter commented:
Do we have a `mooseExperimental` macro or something similar?

dschwen committed:
Add test (#8875)

dschwen commented:
Added a test in 3e5e196

dschwen committed:
Fix test to work in dbg (#8875)

dschwen commented:
Alrighty, this did pass on mac but failed on Linux. The reason seems to be that mac flushes output streams on std::abort while Linux doesn't (and thus drops the output written to Moose::out immediately before std::abort).

permcody commented:
Why are you calling std::abort() though? Shouldn't this be just a normal exit?

We have the _ready_to_exit member Boolean that we normally set. Can you just use that instead?

dschwen commented:
`_ready_to_exit` is protected in MooseApp and not used anywhere outside MooseApp, but if there was a setter for that member this would be an option.

dschwen commented:
@andrsd, test passes now.

andrsd commented:
@dschwen Did you try to override `::solve` with an empty body, so you would not have to call `std::abort` in `initialSetup()`? Just asking...

dschwen commented:
Damn, you're good! I'll try that!

dschwen commented:
Imma take that back. This is a rabbit hole I don't think I want to go down. I get segfaults, `Assertion `this->initialized()' failed.` below `FEProblemBase::advanceState()`, `Assertion `solution->closed()' failed.` below `FEProblemBase::outputStep(MooseEnumItem)` etc. (successively overriding a few more members but is never ending).

andrsd committed:
Using correct base class

andrsd committed:
Fixing API in FEProblemBase

Methods called from Executioners has to be virtual.

andrsd committed:
Dumping the syntax in solve

Do not use std::abort to early terminate the run.  Use the system desing
for cleanly finish.

andrsd commented:
It is not a rabbit hole. It is actually the correct way how to use the class hierarchy of `Problem`s and `System`s. Also, I noticed you were using `FEProblem` as the base class, which I think is not right either, but was easier to do, because otherwise you would need `NonlinearSystemBase`-derived class. Again, I think that is the correct way here. Anyway, check [andrsd/moose:pr/11269](https://github.com/andrsd/moose/tree/pr/11269) for things I suggest should be added to your PR.

dschwen commented:
Thanks for the suggestions @andrsd. I've pulled in your changes, but while the test passes it spits out a nasty memory leak warning now. Let me see if I can get rid of that.

dschwen committed:
Fix memory leak error, formatting, non-unity compile, test (#8875)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/182810/) on 8dac4bf wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11269/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11269/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format df03b0d4f8e48ca193bdffc9243c21cbf73ff1a4`

dschwen commented:
Who would have thought this day would come?
